stages/org.osbuild.copy

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python3
+
+import json
+import os
+import sys
+import subprocess
+
+import osbuild.sources
+
+
+STAGE_DESC = "Copy files from a source to the tree"
+STAGE_INFO = """
+Copies files obtained via a `source` to the tree. Multiple files or
+directories can be copied by specifying multiple entries in `paths`.
+If no paths are specified the whole contents of `source` is copied.
+The source and the target path for each individual path entry might
+optionally be specified via `from` and `to`, respectively; of no
+path is given for any of the two, the root `/` is assumed.
+
+Supported sources are currently:
+
+ o `git`: Fetch the contents of a specific git commit. The commit
+   is specified via its sha1 in the `commit` option. Internally
+   uses the 'org.osbuild.git' source to fetch the commit.
+
+"""
+STAGE_OPTS = """
+"additionalProperties": false,
+"definitions": {
+  "source-git": {
+    "description": "Fetch a checkout via git",
+    "type": "object",
+    "required": ["type", "commit"],
+    "additionalProperties": false,
+    "properties": {
+      "type": {
+        "type": "string",
+         "description": "The type of the source, here 'git'",
+         "enum": ["git"]
+      },
+      "commit": {
+        "type": "string",
+         "description": "The specific commit to fetch"
+      }
+    }
+  }
+},
+"properties": {
+  "source": {
+    "description": "Method of how to obtain the source data",
+    "oneOf": [{ "$ref": "#/definitions/source-git" }]
+  },
```

Comment:
Did you intentionally move the `source-type` into the definition of `source` itself? I can see that it works in JSON as long as you include `type` in every possible option of this `oneOf`. But you basically include the union-identifier inside of the union itself, and relying on every possible union-member to include it at the same path. I would expect `type` to be a top-level identifier and `source` basically your definition but without the type.
Did you intentionally move the type into the union?

Reply:
Yes that was indeed intended to have the schema be definite for the "source" object (can not use `const` because that is in draft 7 and also not `if` because that too is in draft 7. But I agree that this is hard to read. 
I am thinking now that maybe I want separate stages for separate source types, i.e. `org.osbuild.copy.{archive, git, ostree}` and this avoid that.